RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu

```diff
@@ -0,0 +1,498 @@
+
+    for (int i = cellIndexInLayerPair; i < gpuDoublets[layerPairIndex].size;
+         i += gridDim.x * blockDim.x) {
+      auto globalCellIdx = i + globalFirstDoubletIdx;
+
+      auto &thisCell = cells[globalCellIdx];
+      auto innerHitId = thisCell.get_inner_hit_id();
+      auto numberOfPossibleNeighbors =
+          isOuterHitOfCell[globalFirstHitIdx + innerHitId].size();
+      for (auto j = 0; j < numberOfPossibleNeighbors; ++j) {
+        unsigned int otherCell =
+            isOuterHitOfCell[globalFirstHitIdx + innerHitId].m_data[j];
+
+        if (thisCell.check_alignment_and_tag(
+                cells, otherCell, ptmin, region_origin_x, region_origin_y,
+                region_origin_radius, thetaCut, phiCut, hardPtCut)) {
+          //                    printf("kernel_debug_connect: \t\tcell %d is
+          //                    outer neighbor of %d \n", globalCellIdx,
+          //                    otherCell);
+
+          cells[otherCell].theOuterNeighbors.push_back_ts(globalCellIdx);
+        }
+      }
+    }
+  }
+}
+
+template <int maxNumberOfQuadruplets>
+__global__ void kernel_find_ntuplets(
+    unsigned int numberOfRootLayerPairs, const GPULayerDoublets *gpuDoublets,
+    GPUCACell *cells,
+    GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet> *foundNtuplets,
+    unsigned int *rootLayerPairs, unsigned int minHitsPerNtuplet,
+    unsigned int maxNumberOfDoublets) {
+
+  if (blockIdx.y < numberOfRootLayerPairs) {
+    unsigned int cellIndexInRootLayerPair =
+        threadIdx.x + blockIdx.x * blockDim.x;
+    unsigned int rootLayerPairIndex = rootLayerPairs[blockIdx.y];
+    auto globalFirstDoubletIdx = rootLayerPairIndex * maxNumberOfDoublets;
+    GPUSimpleVector<3, unsigned int> stack;
+    for (int i = cellIndexInRootLayerPair;
+         i < gpuDoublets[rootLayerPairIndex].size;
+         i += gridDim.x * blockDim.x) {
+      auto globalCellIdx = i + globalFirstDoubletIdx;
+      stack.reset();
+      stack.push_back(globalCellIdx);
+      cells[globalCellIdx].find_ntuplets(cells, foundNtuplets, stack,
+                                         minHitsPerNtuplet);
+    }
+  }
+}
+
+void CAHitQuadrupletGeneratorGPU::deallocateOnGPU() {
+  cudaStreamDestroy(cudaStream_);
+
+  cudaFreeHost(h_indices);
+  cudaFreeHost(h_doublets);
+  cudaFreeHost(h_x);
+  cudaFreeHost(h_y);
+  cudaFreeHost(h_z);
+  cudaFreeHost(h_rootLayerPairs);
+  for (int i = 0; i < maxNumberOfRegions; ++i)
+    cudaFreeHost(h_foundNtuplets[i]);
+  cudaFreeHost(tmp_layers);
+  cudaFreeHost(tmp_layerDoublets);
+  cudaFreeHost(h_layers);
+
+  cudaFree(d_indices);
+  cudaFree(d_doublets);
+  cudaFree(d_x);
+  cudaFree(d_y);
+  cudaFree(d_z);
+  cudaFree(d_rootLayerPairs);
+  cudaFree(device_theCells);
+  cudaFree(device_isOuterHitOfCell);
+  for (int i = 0; i < maxNumberOfRegions; ++i)
+    cudaFree(d_foundNtuplets[i]);
+}
+
+void CAHitQuadrupletGeneratorGPU::allocateOnGPU() {
+  cudaStreamCreateWithFlags(&cudaStream_, cudaStreamNonBlocking);
```

Comment:
Why do you need to create your own CUDA stream? In addition, after the first call to `hitNtuplets()`, this CUDA stream will not be destroyed in `deallocateOnGPU()`
https://github.com/cms-patatrack/cmssw/blob/d754cfaf488dc6a716afd65c254c644f3e63b4a3/RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cc#L200